In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Define folderpath variable

path = r'C:\Users\Amy Yip\Desktop\Career Foundry Data Analytics\Achievement 4 Python\05-02-2022 Instacart Basket Analysis\02 Data'

In [3]:
# Import ords_prods_merged (inner join) dataframe

df_ords_prods_merged = pd.read_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merged.pkl'))

In [4]:
df_ords_prods_merged.head()

order_id  user_id  order_numbers_users_made  orders_day_of_week  \
0   2539329        1                         1                   2   
1   2398795        1                         2                   3   
2    473747        1                         3                   3   
3   2254736        1                         4                   4   
4    431534        1                         5                   4   

   order_hour_of_day  days_since_prior_order  prior_orders_made  product_id  \
0                  8                     NaN               True         196   
1                  7                    15.0              False         196   
2                 12                    21.0              False         196   
3                  7                    29.0              False         196   
4                 15                    28.0              False         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                  1          0         Soda        77              7     9.0   
1                  1          1         Soda        77              7     9.0   
2                  1          1         Soda        77              7     9.0   
3                  1          1         Soda        77              7     9.0   
4                  1          1         Soda        77              7     9.0   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

In [5]:
df_ords_prods_merged.shape

(32404859, 15)

In [6]:
# Create a subset of the first one million rows using colon

df = df_ords_prods_merged[:1000000]

In [7]:
df.shape

(1000000, 15)

In [8]:
df.head()

order_id  user_id  order_numbers_users_made  orders_day_of_week  \
0   2539329        1                         1                   2   
1   2398795        1                         2                   3   
2    473747        1                         3                   3   
3   2254736        1                         4                   4   
4    431534        1                         5                   4   

   order_hour_of_day  days_since_prior_order  prior_orders_made  product_id  \
0                  8                     NaN               True         196   
1                  7                    15.0              False         196   
2                 12                    21.0              False         196   
3                  7                    29.0              False         196   
4                 15                    28.0              False         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                  1          0         Soda        77              7     9.0   
1                  1          1         Soda        77              7     9.0   
2                  1          1         Soda        77              7     9.0   
3                  1          1         Soda        77              7     9.0   
4                  1          1         Soda        77              7     9.0   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

#### 1. If you haven’t done so already, complete the instructions in the Exercise for creating the “price_label” and “busiest_day” columns.

#### (i) 'price_label' column - Using User-Defined Function

In [9]:
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range product'
  else: return 'Not enough data'

In [10]:
# Use the dataframe to specify the appropriate string label to return for each product.

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\AMYYIP~1\AppData\Local\Temp/ipykernel_18572/3955797688.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [11]:
df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [12]:
# Check the most expensive product within the subset

df['prices'].max()

14.8

In [16]:
# Alternative: using If-Statements with loc() Function

df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [14]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [15]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [17]:
df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

In [18]:
# Using loc() on the entire dataframe

df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [19]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [20]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [21]:
df_ords_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

#### (ii) 'Busiest day' column - Using If-Statements with For Loops

In [22]:
# Frequency of the “orders_day_of_week” column

df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [23]:
# Using for-loop to create new labels for busiest day 

result = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [24]:
# Print result

result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [25]:
# Create new column 'busiest_day' to contain these labels

df_ords_prods_merged['busiest_day'] = result

In [26]:
# Value counts for each label

df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

#### 2. Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [27]:
# Using for-loop to create new labels for busiest days 

result2 = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result2.append("Busiest days")
  elif value == 1:
    result2.append("Busiest days")
  elif value == 4:
    result2.append("Slowest days")
  elif value == 3:
    result2.append("Slowest days")
  else:
    result2.append("Regularly busy")

In [28]:
# Create new column 'busiest_days' to contain these labels

df_ords_prods_merged['busiest_days'] = result2

In [29]:
# Value counts for each label

df_ords_prods_merged['busiest_days'].value_counts(dropna = False)

Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: busiest_days, dtype: int64

#### 3. Check the values of this new column for accuracy. Note any observations in markdown format.

* Sum of orders on 0 and 1 = 6204182 + 5660230 = 11864412
* Sum of orders on 3 and 4 = 3840534 + 3783802 = 7624336

#### Previous answer using left join:
* Sum of orders on 0 and 1 = 6252941 + 5696724 = 11949665
* Sum of orders on 3 and 4 = 3868747 + 3812390 = 7681137

#### 4. When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

#### 5. Print the frequency for this new column.

In [30]:
# Identify busiest hour - by determining the frequency of the “orders_day_of_week” column

df_ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

10am is the busiest hour of the day, while 3 am is the least busiest hour of the day.

For the following exercise, I will classify 10-19th hour as the busiest, while 2-5th hour as the least busiest.

In [31]:
# Using for-loop to create new labels for the time period of orders 

result_hour = []

for value in df_ords_prods_merged["order_hour_of_day"]:
  if value >= 10 & value <= 19:
    result_hour.append("Most orders")
  elif value >= 2 & value <= 5:
    result_hour.append("Fewest orders")
  else:
    result_hour.append("Average orders")

In [32]:
# Print result_hour

result_hour

['Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most o

In [33]:
# Create a new column containing these labels called “busiest_period_of_day.”

df_ords_prods_merged['busiest_period_of_day'] = result_hour

In [34]:
df_ords_prods_merged.head()

order_id  user_id  order_numbers_users_made  orders_day_of_week  \
0   2539329        1                         1                   2   
1   2398795        1                         2                   3   
2    473747        1                         3                   3   
3   2254736        1                         4                   4   
4    431534        1                         5                   4   

   order_hour_of_day  days_since_prior_order  prior_orders_made  product_id  \
0                  8                     NaN               True         196   
1                  7                    15.0              False         196   
2                 12                    21.0              False         196   
3                  7                    29.0              False         196   
4                 15                    28.0              False         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                  1          0         Soda        77              7     9.0   
1                  1          1         Soda        77              7     9.0   
2                  1          1         Soda        77              7     9.0   
3                  1          1         Soda        77              7     9.0   
4                  1          1         Soda        77              7     9.0   

  _merge    price_range_loc     busiest_day    busiest_days  \
0   both  Mid-range product  Regularly busy  Regularly busy   
1   both  Mid-range product  Regularly busy    Slowest days   
2   both  Mid-range product  Regularly busy    Slowest days   
3   both  Mid-range product      Least busy    Slowest days   
4   both  Mid-range product      Least busy    Slowest days   

  busiest_period_of_day  
0           Most orders  
1           Most orders  
2           Most orders  
3           Most orders  
4           Most orders

In [35]:
df_ords_prods_merged.shape

(32404859, 19)

In [36]:
# Dropping columns: '_merge', 'busiest_day'
# '_merge' is not required as we are using inner join, so the result will all be 'both'.
# 'busiest_day' is no longer required as clients have changed their minds “busiest_day” column.

df_ords_prods_merged = df_ords_prods_merged.drop(columns = ['_merge', 'busiest_day'])

In [37]:
df_ords_prods_merged.head()

order_id  user_id  order_numbers_users_made  orders_day_of_week  \
0   2539329        1                         1                   2   
1   2398795        1                         2                   3   
2    473747        1                         3                   3   
3   2254736        1                         4                   4   
4    431534        1                         5                   4   

   order_hour_of_day  days_since_prior_order  prior_orders_made  product_id  \
0                  8                     NaN               True         196   
1                  7                    15.0              False         196   
2                 12                    21.0              False         196   
3                  7                    29.0              False         196   
4                 15                    28.0              False         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                  1          0         Soda        77              7     9.0   
1                  1          1         Soda        77              7     9.0   
2                  1          1         Soda        77              7     9.0   
3                  1          1         Soda        77              7     9.0   
4                  1          1         Soda        77              7     9.0   

     price_range_loc    busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy           Most orders  
1  Mid-range product    Slowest days           Most orders  
2  Mid-range product    Slowest days           Most orders  
3  Mid-range product    Slowest days           Most orders  
4  Mid-range product    Slowest days           Most orders

In [38]:
df_ords_prods_merged.shape

(32404859, 17)

#### 6. Ensure your notebook is clean and structured and that your code is well commented.

#### 7. Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.

In [39]:
df_ords_prods_merged.to_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merged_new_cols.pkl'))

#### 8. Save your notebook and submit it to your tutor for review.